File: services/frontend/src/app/core/store/inventory.repository.ts
/*
 * G4IT
 * Copyright 2023 Sopra Steria
 *
 * This product includes software developed by
 * French Ecological Ministery (https://gitlab-forge.din.developpement-durable.gouv.fr/pub/numeco/m4g/numecoeval)
 */ 
import { Injectable } from "@angular/core";
import { createStore, select, setProp, withProps } from "@ngneat/elf";
import { selectAllEntities, setEntities, withEntities } from "@ngneat/elf-entities";
import { Inventory } from "src/app/core/interfaces/inventory.interfaces";

export interface InventoryProps {
    selected?: string;
}

const inventoryStore = createStore(
    { name: "Inventory" },
    withProps<InventoryProps>({ selected: "" }),
    withEntities<Inventory, "id">()
);

@Injectable({ providedIn: "root" })
export class InventoryRepository {
    inventories$ = inventoryStore.pipe(selectAllEntities());
    selectedInventory$ = inventoryStore.pipe(select((state) => state.selected));

    updateSelectedInventory(id: any) {
        inventoryStore.update(setProp("selected", id));
    }

    setInventories(inventories: Inventory[]) {
        const inventoriesWithIds = inventories.map((inventory) => ({
            ...inventory,
            id: inventory.id,
        }));
        inventoryStore.update(setEntities(inventoriesWithIds));
    }
}

File: services/backend/src/main/java/com/soprasteria/g4it/backend/apiinventory/repository/InventoryRepository.java
/*
 * G4IT
 * Copyright 2023 Sopra Steria
 *
 * This product includes software developed by
 * French Ecological Ministery (https://gitlab-forge.din.developpement-durable.gouv.fr/pub/numeco/m4g/numecoeval)
 */
package com.soprasteria.g4it.backend.apiinventory.repository;

import com.soprasteria.g4it.backend.apiinventory.modeldb.Inventory;
import com.soprasteria.g4it.backend.apiuser.modeldb.Workspace;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Modifying;
import org.springframework.data.jpa.repository.Query;
import org.springframework.data.repository.query.Param;
import org.springframework.stereotype.Repository;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

/**
 * Inventory Repository to access Inventory Data in database.
 */
@Repository
public interface InventoryRepository extends JpaRepository<Inventory, Long> {

    /**
     * Delete inventory and return number of lines deleted
     *
     * @param inventoryId the unique inventory identifier
     */
    @Transactional
    @Modifying
    @Query("delete from Inventory i where i.id=:inventoryId")
    void deleteByInventoryId(@Param("inventoryId") final Long inventoryId);

    /**
     * Find by organization and the inventory id and return  a list of matching inventories
     *
     * @param workspace the unique workspace identifier.
     * @return a list of matching inventories
     */
    Optional<Inventory> findByWorkspaceAndId(final Workspace workspace,
                                             final Long inventoryId);

    /**
     * Find by workspace and workspace then return a list of inventories
     *
     * @param workspace the linked workspace
     * @return a list of inventory
     */
    List<Inventory> findByWorkspace(final Workspace workspace);


    /**
     * Find by workspace and the inventory name and return  a list of matching inventories
     *
     * @param workspace the unique workspace identifier.
     * @param name      the inventory name.
     * @return a list of matching inventories
     */
    Optional<Inventory> findByWorkspaceAndName(final Workspace workspace, final String name);

    /**
     * Update the output counts of an inventory without loading/saving the full entity graph.
     * <p>
     * NB: {@code Inventory.tasks} is mapped as EAGER + CascadeType.ALL, so using
     * {@code inventoryRepository.save(inventory)} here would cascade-save any stale
     * {@code Task} entities still referenced in memory (e.g. loaded before an evaluation
     * started), overwriting concurrent progress updates made via bulk queries
     * (see {@code TaskRepository.updateProgress}). This targeted update avoids that.
     *
     * @param inventoryId          the unique inventory identifier
     * @param outPhysicalCount     the output physical equipment count
     * @param outVirtualCount      the output virtual equipment count
     * @param outApplicationCount  the output application count
     */
    @Transactional
    @Modifying
    @Query("""
            update Inventory i
            set i.outPhysicalCount = :outPhysicalCount,
                i.outVirtualCount = :outVirtualCount,
                i.outApplicationCount = :outApplicationCount
            where i.id = :inventoryId
            """)
    void updateOutputCounts(@Param("inventoryId") final Long inventoryId,
                            @Param("outPhysicalCount") final Long outPhysicalCount,
                            @Param("outVirtualCount") final Long outVirtualCount,
                            @Param("outApplicationCount") final Long outApplicationCount);

}
